### PR TITLE
NET-655

### PR DESCRIPTION
**Pull request by uGiFarukh**
## Describe your changes
* When MQ connection is broken, a fallback thread is being ran every 30s for calling netclient http config pull.
* If peer updates are not received or lost, the same fallback thread will run every 30s for calling netclient http config pull.

Related to -> https://github.com/gravitl/netmaker/pull/2670

## Checklist before requesting a review
- [x] My changes affect only 10 files or less.
- [x] I have performed a self-review of my code and tested it.
- [x] If it is a new feature, I have added thorough tests, my code is <= 1450 lines.
- [ ] If it is a bugfix, my code is <= 200 lines.
- [x] My functions are <= 80 lines.
- [x] I have had my code reviewed by a peer.
- [x] My unit tests pass locally.
- [x] Netclient & Netmaker are awesome.

### CODE REVIEW

File: functions/mqhandlers.go [hunk 1]
```diff
@@ -573,3 +590,25 @@ func handleFwUpdate(server string, payload *models.FwUpdate) {
...
+		default:
+			MQFallbackRunning = true
+
+			// Call http netclient config pull
+			_, err := Pull(false)
+			if err != nil {
+				slog.Error("failed to pull")
+			}
+
+			time.Sleep(time.Second * 30) // Execute pull every 30 seconds
```
Comment (mattkasun) [inline functions/mqhandlers.go:611]: use a time.Ticker

**Review by mattkasun — CHANGES_REQUESTED** [review]
would it not be simpler to start the fallback go routine in the mq connection lost handler and stop it in the mq connection handler

File: functions/mqhandlers.go (hunk 1 above)
Comment (uGiFarukh) [reply functions/mqhandlers.go:611]: I thought so. But @abhishek9686 wants this to handle both mq connection dropout situation and mq peer updates lost situation.

File: functions/daemon.go [hunk 2]
```diff
@@ -109,6 +109,13 @@ func closeRoutines(closers []context.CancelFunc, wg *sync.WaitGroup) {
 	slog.Info("closing netmaker interface")
 	iface := wireguard.GetInterface()
 	iface.Close()
+
+	// Closing MQTT Fallback Go Routines
+	MQFallbackTickerStop <- true // signal the tick based goroutine to stop
+	if MQFallbackRunning {
```
Comment (abhishek9686) [inline functions/daemon.go:115]: use lock or sync/atomic for accessing `MQFallbackRunning`

File: functions/daemon.go [hunk 3]
```diff
@@ -187,6 +194,30 @@ func startGoRoutines(wg *sync.WaitGroup) context.CancelFunc {
 	go Checkin(ctx, wg)
 	wg.Add(1)
 	go networking.StartIfaceDetection(ctx, wg, config.Netclient().ListenPort)
+
+	// MQTT Fallback Channel and Ticker
+	MQFallbackStop = make(chan bool)
+	MQFallbackTicker = time.NewTicker(MQ_FALLBACK_TIMEOUT)
+	MQFallbackTickerStop = make(chan bool)
+
```
Comment (abhishek9686) [inline functions/daemon.go:202]: these vars can defined where there are declared

File: functions/daemon.go [hunk 4]
```diff
@@ -187,6 +194,30 @@ func startGoRoutines(wg *sync.WaitGroup) context.CancelFunc {
...
+	// MQTT Peer Update fallback ticker
+	go func() {
```
Comment (abhishek9686) [inline functions/daemon.go:204]: move this into a function and use the wait group to exit safely when main thread stops

File: functions/mqhandlers.go [hunk 5]
```diff
@@ -573,3 +590,25 @@ func handleFwUpdate(server string, payload *models.FwUpdate) {
...
+// MQTT Fallback Mechanism
+func MQFallback(MQFallbackStop chan bool) {
+	for {
+		select {
+		case <-MQFallbackStop:
+			// Stop the goroutine
+			MQFallbackRunning = false
+			return
+		default:
+			MQFallbackRunning = true
```
Comment (abhishek9686) [inline functions/mqhandlers.go:603]: do we need to set this `variable` here on every loop?

File: functions/daemon.go
Comment (abhishek9686) [inline functions/daemon.go:1]: @uGiFarukh the main place you need invoke the fall back thread is when Mq connection fails to establish, and I don't see that's been covered in this PR

**Review by abhishek9686 — CHANGES_REQUESTED** [review]

File: functions/daemon.go [hunk 6]
```diff
@@ -198,6 +211,12 @@ func messageQueue(ctx context.Context, wg *sync.WaitGroup, server *config.Server
 	err := setupMQTT(server)
 	if err != nil {
 		slog.Error("unable to connect to broker", "server", server.Broker, "error", err)
+
+		// start mqtt fallback goroutine if it is not running
+		if !MQFallbackRunning.Load() {
```
Comment (abhishek9686) [inline functions/daemon.go:216]: if you are using these variable within the same package then you can keep them private

File: functions/mqhandlers.go [hunk 7]
```diff
@@ -26,6 +27,16 @@ import (
...
+// MQTT Fallback Routine
+var MQFallbackRunning atomic.Bool
+var MQFallbackStop chan bool = make(chan bool)
+var MQFallbackTicker *time.Ticker = time.NewTicker(MQ_FALLBACK_TICK)
+var MQMessageLostFallbackTickerStop chan bool = make(chan bool)
+var MQMessageLostFallbackTicker *time.Ticker = time.NewTicker(MQ_FALLBACK_TIMEOUT)
+
+const MQ_FALLBACK_TICK = time.Second * 30
+const MQ_FALLBACK_TIMEOUT = time.Minute * 6
+
```
Comment (abhishek9686) [inline functions/mqhandlers.go:39]: you can make these vars private

File: functions/daemon.go [hunk 8]
```diff
@@ -198,6 +211,12 @@ func messageQueue(ctx context.Context, wg *sync.WaitGroup, server *config.Server
 	err := setupMQTT(server)
 	if err != nil {
 		slog.Error("unable to connect to broker", "server", server.Broker, "error", err)
+
```
Comment (abhishek9686) [inline functions/daemon.go:214]: also, it will be safe to check on a successful connection to mq to stop the fallback thread if it's running already

**Review by abhishek9686 — CHANGES_REQUESTED** [review]

**Review by mattkasun — CHANGES_REQUESTED** [review]
I think this is way over complicated with starting stopping the fallback function all over the place:

just need to start fallback go routine and let it run

```

func MQFallback(ctx contex.Context, wg *sync.WaitGroup) {
	defer wg.Done()
	for {
		select {
		case <-ctx.Done():
			fallbackTicker.Stop()
			return
		case <-fallbackTicker.C:
			if Mqclient == nil || !Mqclient.IsConnected() {
				response, err := pull(false)
				if err != nil {
					slog.Error("pull failed", "error", err)
				} else {
					MQFallbackPull(response)
				}
			}
		}
	}
}

```

File: functions/mqhandlers.go
Comment (abhishek9686) [inline functions/mqhandlers.go:1]: what about the code to check if there any new nodes added?
then you have to restart the daemon

**Review by abhishek9686 — CHANGES_REQUESTED** [review]

File: functions/mqhandlers.go [hunk 9]
```diff
@@ -573,3 +578,86 @@ func handleFwUpdate(server string, payload *models.FwUpdate) {
 	}
 
 }
+
+// MQTT Fallback Mechanism
+func MQFallback(ctx context.Context, wg *sync.WaitGroup) {
```
Comment (abhishek9686) [inline functions/mqhandlers.go:583]: funcs in same package doesn't have to be exported

**Review by abhishek9686 — CHANGES_REQUESTED** [review]

**Review by abhishek9686 — APPROVED** [review]

**Pull request merged**